### ISSUE TITLE
why complain about prototype not in roots?

### ISSUE THREAD

**Issue by dckc** [Agoric/harden#54]
What's the purpose / justification of this test?

https://github.com/Agoric/harden/blob/e29849c18e0cfbbd41db1e8765954d4bc51d5e4a/test/test.js#L14

This is the only test that fails when I implement `harden` using [deep freezing](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md#deep-freezing) from the moddable SDK:

```
export default function harden(root) {
  return Object.freeze(root, true);
}
```

cc @erights @warner @phoddie

**Comment by phoddie** [comment]
Yes, XS extends `Object.freeze` with an optional second argument to recursively freeze the first argument. That is [described](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md#deep-freezing) in the preload documentation.

**Comment by warner** [comment]
I'll let @erights confirm, but I think our concern is that an object's inheritance chain is also part of its external API surface, so we don't want to expose supposedly-hardened objects which still have mutable prototypes. If all API-visible objects inherit from `Object` (which is a frozen primordial) or have a `null` prototype, then we're good. But if a user creates a new pair of objects, makes A inherit from B, freezes A but not B, then exposes A, now an attacker can add arbitrary (non-`own`) properties to A by modifying B.

So our rule for `harden` is that all enumerable properties, *plus the `.prototype`, if any*, must also be hardened. And the definition of "hardened" is that it is present in the "root set".

Does XS's `Object.freeze(x, true)` also freeze the `.prototype` ?

**Comment by phoddie** [comment]
> Does XS's Object.freeze(x, true) also freeze the .prototype ?

The following code fails with `set x: not extensible!`.

```js
class Foo {
	test() {
	}
}
Object.freeze(Foo, true);

Foo.prototype.x = true;
```

**Comment by dckc** [comment]
> ... And the definition of "hardened" is that it is present in the "root set".

That seems overly constraining.

I'm not sure how to move the ball forward here. @erights can you think of a way to refine the test?

**Comment by erights** [comment]
Expanding a bit on what @warner says at https://github.com/Agoric/harden/issues/54#issuecomment-533322212

I need to explain the rationale in depth when I have the time. The short form is that harden and deep freeze serve different purposes. Harden originally did transitively walk both own properties and [[Prototype]], freezing everything in the transitive closure. For use of harden at runtime, that was too infectious and hard to reason about. Rather, we decided that harden should ensure that hardened things only inherit from hardened things (modulo a list of exceptions). This means examining [[Prototype]]s non-transitively, not harden by those paths, but reject if the harden invariant would be violated.

This means that for some hardens to succeed, other hardens, or things higher up in the inheritance chain, must be hardened first. So far this has not been a problem. I don't understand how we're running into that problem here. If Moddable's two-argument deep `freeze` is more aggressive and done first, I don't see how this would cause a `harden` to fail that, without the deep `freeze`, would have succeeded.

**Comment by erights** [comment]
At https://github.com/Agoric/harden/issues/54#issue-488334111 @dckc writes:

> when I implement harden using deep freezing from the moddable SDK
>```js
> export default function harden(root) {
>   return Object.freeze(root, true);
> }
```

Whether on not this actually causes tests to fail, it is not correct. By design, Moddable's deep `freeze` is more aggressive than `harden`. For cases where both succeed, IIUC, they are essentially equivalent. However, `harden` is designed to fail for some cases --- those encountered above --- where deep `freeze` is designed to succeed by freezing more.

**Comment by erights** [comment]
@phoddie,
https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md#deep-freezing says

>  If equivalent functionality becomes available in a standard way such as `harden`, XS will move to use that mechanism exclusively.

We should definitely revisit and discuss to ensure we have a set of hardening/freezing/snapshotting operations that fulfill all of our needs. `harden` alone may not be sufficient. But if more is needed, it should also become a standard part of EcmaScript.

**Comment by dckc** [comment]
@erights about harden and tape-xs... this is the blocker.

In discussion with @jfparadis , he concurred that the whole business of tracing frozen stuff with a WeakMap is kinda silly in xs where so much stuff is frozen to start with. In particular, this optimization of not hardening prototypes is a noop on xs, since the prototypes will already be frozen. This test seems to be testing an implementation detail, not a feature.
